packages/api/test/analytics-test.js

```diff
@@ -0,0 +1,185 @@
+
+      // buildata not changed
+      const buildData = await client.collection(global.headCollection).doc(buildInfo.repoId).collection('builds').doc('0').get();
+      assert.strictEqual(buildData.data().flaky, 0);
+
+      // test case flakyness not changed
+      const testFlakyExpecations = [false, true, true, false];
+      for (let i = 0; i < testFlakyExpecations.length; i++) {
+        const testData = await client.collection(global.headCollection).doc(buildInfo.repoId).collection('tests').doc(i.toString()).get();
+        assert.strictEqual(testData.data().flaky, testFlakyExpecations[i]);
+      }
+    });
+
+    it('Can add builds non chronologically', async () => {
+      const updateObj = JSON.parse(JSON.stringify(buildInfo));
+      updateObj.timestamp = new Date('06/01/2021');
+      updateObj.buildId = '1.5';
+      const testCaseObjs = [];
+      const nonChronTestCases = [true, false, false, true];
+      for (let k = 0; k < nonChronTestCases.length; k++) {
+        testCaseObjs.push(new TestCaseRun(nonChronTestCases[k] ? 'ok' : 'not ok', k, k.toString()));
+      }
+
+      // readding the first build should mantain its status as nonflaky
+      // since only looking at test cases with timestamp
+      await addBuild(testCaseObjs, updateObj, client, global.headCollection);
+
+      // check builddata
+      const buildData = await client.collection(global.headCollection).doc(buildInfo.repoId).collection('builds').doc('1.5').get();
+      assert.strictEqual(buildData.data().flaky, 1); // only 1 is flaky at this point
+
+      // ensure test cases flayness is unchanged
+      const testFlakyExpecations = [false, true, true, true]; // note test case 4 has now become flaky
+      for (let i = 0; i < testFlakyExpecations.length; i++) {
+        const testData = await client.collection(global.headCollection).doc(buildInfo.repoId).collection('tests').doc(i.toString()).get();
+        assert.strictEqual(testData.data().flaky, testFlakyExpecations[i]);
+      }
+
+      // repo info changes flaky count, does not change num test cases/ numfails count
+      const repoData = await client.collection(global.headCollection).doc(buildInfo.repoId).get();
+      assert.strictEqual(repoData.data().numfails, 3); // number of tests/failing testss has not changed, since this is not up to date
+      assert.strictEqual(repoData.data().numtestcases, 4);
+      assert.strictEqual(repoData.data().flaky, 3); // # flaky tests has Changed
+    });
+  });
+
+  after(async () => {
+    // must delete each collection individually
+    const deletePaths = [
+      'tests/{testcase}/runs/{buildid}',
+      'tests/{testcase}',
+      'builds/{buildid}'
+    ];
+    const buildIds = ['1', '2', '3', '4'];
+    const testCaseIds = [];
+    for (let j = 0; j < testCases.length; j++) {
+      testCaseIds.push(j.toString());
+    }
+    // Delete all possible documents, okay to delete document that doesnt exist
+    deletePaths.forEach(async (deletePath) => {
```

Comment:
Mixing `forEach` and `async` can be a bit risky, because promises eventual values. I think this will end up creating N promises immediately, not delete test data in sequence.

If you do want to delete several items in parallel, it can be more idiomatic to go:

```
await Promise.all(testCaseIds.map(async () => {
     const deletePathUse = deletePath.replace('{testcase}', encodeURIComponent(testCase)).replace('{buildid}', buildId);
      await client.collection(global.headCollection).doc(buildInfo.repoId + '/' + deletePathUse).delete();
});
```

If you want to delete in sequence, you might do better to unravel this into a for statement.